src/cargo/util/config/mod.rs

```diff
@@ -483,6 +483,23 @@ impl Config {
         Ok(())
     }
 
+    /// Updates the application's notion of current working directory, both in
+    /// the process environment as well as the configuration data
+    pub fn set_cwd(&mut self, new_cwd: &Path) -> CargoResult<()> {
+        // Update the process-level notion of cwd
+        std::env::set_current_dir(&new_cwd)?;
```

Comment:
Feels a bit dirty to be modifying a global inside of a function to modify a global.  I wonder if there is another way for us to do this?

Reply:
This was included for a couple reasons, but perhaps not necessary.

Cargo stashes the current directory in it's config data, and _almost_ everything looks there to know what the current directory is.

After auditing for code that checks `current_dir()` instead of Cargo's config data, I found a few call sites that don't or can't use the value from Config:

- cargo/ops/cargo_add/mod.rs: `resolve_dependency()` normalizes paths relative to actual cwd, instead of the Config value
- cargo/core/package_id_spec.rs: the `parse()` function assumes non-URI path types are paths relative to actual cwd
- cargo/core/compiler/timings.rs: `report_html()` saves the report relative to actual cwd

Cargo also launches subprocess, and it sets an appropriate cwd for _almost_ all of them.

After auditing subprocess calls, I found the following that don't set an appropriate cwd on called processes:

- src/cargo/ops/registry/auth.rs: all processes launched by run_command() will inherit the actual cwd of cargo, instead of the Config value
- src/cargo/ops/cargo_doc.rs: the launched browser inherits the actual cwd of cargo, this _seems_ benign

Ideally we wouldn't mess with the stored Config value at all, and we'd just `std::env::set_current_dir(new_cwd)` _before_ it gets read for being stored away in Config, but the cli args aren't ready then.

So as for doing this another way, I'm uncertain how to keep the two values in sync otherwise.

Reply:
I'm in the process of re-validating the current code after re-basing on top of the tip of master.  After I know I've got that working again, I'll try your suggestion out in another branch.